### PR TITLE
Change ClusterIP in dns e2e test

### PR DESCRIPTION
**Pull request by xiangpengzhao**
**What this PR does / why we need it**:
Since we may not want to cherry-pick #46197, I create this PR to make a small patch which can be cherry-picked to 1.6/1.7 to deflake #50274.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #
ref:
https://github.com/kubernetes/kubernetes/issues/50274#issuecomment-320862204
https://github.com/kubernetes/kubernetes/issues/50274#issuecomment-321733398

**Special notes for your reviewer**:
/cc @thockin @ericchiang @apelisse @MrHohn 

**Release note**:

```release-note
NONE
```

### CODE REVIEW

**Comment by thockin** [comment]
I don't understand the bug that this is fixing or why it is flaky as opposed to a full failure.  Can you explain in the PR comment, please?

**Comment by xiangpengzhao** [comment]
@thockin this PR itself doesn't fix anything, but after it is cherry-picked to 1.6/17, it will help pass the validation of IP range for the dns e2e tests of those releases. Currently the dns e2e test `DNS should provide DNS for ExternalName services` specifies ClusterIP to be `127.1.2.3` which is an invalid ClusterIP. 

Since I can't send a PR to fix the ClusterIP in 1.6/1.7 directly, I just create this one so that I can have a PR to cherry-pick.

**Comment by xiangpengzhao** [comment]
Even though @MrHohn told me the upgrade workflow in https://github.com/kubernetes/kubernetes/issues/51695#issuecomment-326430304, I'm still a bit confused by the upgrade test log here:
https://k8s-gubernator.appspot.com/build/kubernetes-jenkins/logs/ci-kubernetes-e2e-gke-gci-new-gci-master-upgrade-master/467?log#log.

@MrHohn said that
```
The expected sequence for this specific cluster upgrade job is:

1. Create a 1.7 cluster.
2. Start a bunch of upgrade tests (initiation phase).
3. Upgrade the master.
4. Upgrade the nodes.
5. Finish those upgrade tests (evaluation phase).
6. Run skew tests (a set of tests on latest branch) against the upgraded cluster.
```

But in the log, what I see is that seems like 1.7 test cases are run against a master branch cluster.
1. master is 1.7 before upgrade.
```
W0919 20:27:34.700] 2017/09/19 20:27:34 util.go:154: Running: ./hack/e2e-internal/e2e-status.sh
W0919 20:27:34.743] Skeleton Provider: prepare-e2e not implemented
I0919 20:27:34.874] Client Version: version.Info{Major:"1", Minor:"9+", GitVersion:"v1.9.0-alpha.0.741+a238fbd2539add", GitCommit:"a238fbd2539addc525bb740ae801a42524e5b706", GitTreeState:"clean", BuildDate:"2017-09-19T18:40:31Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
I0919 20:27:34.874] Server Version: version.Info{Major:"1", Minor:"7+", GitVersion:"v1.7.7-beta.0.12+d77ea306f434d3", GitCommit:"d77ea306f434d35094896bcd3d195a21436f4a62", GitTreeState:"clean", BuildDate:"2017-09-17T04:40:00Z", GoVersion:"go1.8.3", Compiler:"gc", Platform:"linux/amd64"}
W0919 20:27:34.975] 2017/09/19 20:27:34 util.go:156: Step './hack/e2e-internal/e2e-status.sh' finished in 178.47657ms
W0919 20:27:34.976] 2017/09/19 20:27:34 util.go:154: Running: ./hack/ginkgo-e2e.sh --ginkgo.focus=\[Feature:MasterUpgrade\] --upgrade-target=ci/latest --upgrade-image=gci --num-nodes=3 --report-dir=/workspace/_artifacts --disable-log-dump=true --report-prefix=upgrade
```
2. after upgrade, the master is 1.9
```
I0919 20:38:00.568] Sep 19 20:38:00.567: INFO: Upgrading master to "1.9.0-alpha.0.741+a238fbd2539add"
I0919 20:38:00.568] Sep 19 20:38:00.567: INFO: Running gcloud [container clusters --project=gke-up-c1-3-c1-5-up-clu-n --zone=us-central1-a upgrade bootstrap-e2e --master --cluster-version=1.9.0-alpha.0.741+a238fbd2539add --quiet]

... ...

I0919 20:42:09.543] Sep 19 20:42:09.542: INFO: Checking master version
I0919 20:42:09.551] Sep 19 20:42:09.551: INFO: Master is at version 1.9.0-alpha.0.741+a238fbd2539add

... ...

I0919 20:57:01.917] • [SLOW TEST:1752.843 seconds]
I0919 20:57:01.917] [sig-cluster-lifecycle] Upgrade [Feature:Upgrade]
I0919 20:57:01.917] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/lifecycle/framework.go:22
I0919 20:57:01.917]   master upgrade
I0919 20:57:01.918]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/lifecycle/cluster_upgrade.go:101
I0919 20:57:01.918]     should maintain a functioning cluster [Feature:MasterUpgrade]
I0919 20:57:01.918]     /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/lifecycle/cluster_upgrade.go:100
I0919 20:57:01.918] ------------------------------
```
3. Then the test case `should provide DNS for ExternalName services` failed.
```
I0920 02:09:49.831] • Failure [31.753 seconds]
I0920 02:09:49.831] [k8s.io] DNS
I0920 02:09:49.831] /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:619
I0920 02:09:49.832]   should provide DNS for ExternalName services [It]
I0920 02:09:49.832]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/dns.go:488
I0920 02:09:49.832] 
I0920 02:09:49.832]   Expected error:
I0920 02:09:49.833]       <*errors.StatusError | 0xc420a63480>: {
I0920 02:09:49.833]           ErrStatus: {
I0920 02:09:49.833]               TypeMeta: {Kind: "", APIVersion: ""},
I0920 02:09:49.833]               ListMeta: {SelfLink: "", ResourceVersion: ""},
I0920 02:09:49.834]               Status: "Failure",
I0920 02:09:49.834]               Message: "Service \"dns-test-service-3\" is invalid: spec.clusterIP: Invalid value: \"127.1.2.3\": provided IP is not in the valid range",
I0920 02:09:49.834]               Reason: "Invalid",
I0920 02:09:49.834]               Details: {
I0920 02:09:49.835]                   Name: "dns-test-service-3",
I0920 02:09:49.835]                   Group: "",
I0920 02:09:49.835]                   Kind: "Service",
I0920 02:09:49.835]                   UID: "",
I0920 02:09:49.835]                   Causes: [
I0920 02:09:49.836]                       {
I0920 02:09:49.836]                           Type: "FieldValueInvalid",
I0920 02:09:49.836]                           Message: "Invalid value: \"127.1.2.3\": provided IP is not in the valid range",
I0920 02:09:49.836]                           Field: "spec.clusterIP",
I0920 02:09:49.836]                       },
I0920 02:09:49.837]                   ],
I0920 02:09:49.837]                   RetryAfterSeconds: 0,
I0920 02:09:49.837]               },
I0920 02:09:49.837]               Code: 422,
I0920 02:09:49.837]           },
I0920 02:09:49.837]       }
I0920 02:09:49.838]       Service "dns-test-service-3" is invalid: spec.clusterIP: Invalid value: "127.1.2.3": provided IP is not in the valid range
I0920 02:09:49.838]   not to have occurred
I0920 02:09:49.838] 
I0920 02:09:49.838]   /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/dns.go:477
```

I didn't find what the e2e test code version is. But according to the code path `test/e2e/dns.go` and the ClusterIP shown in the log, the e2e test code version seems like to be 1.7.

1.7 code
https://github.com/kubernetes/kubernetes/blob/release-1.7/test/e2e/dns.go#L472
```go
s.Spec.ClusterIP = "127.1.2.3"
```

master code
https://github.com/kubernetes/kubernetes/blob/master/test/e2e/network/dns.go#L452
```go
s.Spec.ClusterIP = "10.0.0.123"
```

So, this means that a 1.7 test case `should provide DNS for ExternalName services` running against a 1.9(master branch) cluster. Then the test WILL fail ! It's because #46197 (merged after 1.7) added the validation of ClusterIP when updating service type.

Hope I have explained clearly what happened...

Is running a 1.7 test case against a master branch cluster expected in the workflow of 1.7-master upgrade test?

**Comment by xiangpengzhao** [comment]
So I want to change the ClusterIP (`s.Spec.ClusterIP = "127.1.2.3"`) in the test case of 1.7 version to be a valid one, e.g, `10.0.0.111` as is shown in this PR. Then we don't need to care about which release test case running against which cluster. The validation will always pass.

**Comment by jdumars** [comment]
Added to 1.8 milestone, but still needs the regular approval LGTM/Approve

**Comment by MrHohn** [comment]
> Is running 1.7 test cases against a master branch cluster expected in the workflow of 1.7-master upgrade test?

@xiangpengzhao Yes, we run test cases in latest branch and in 1.7 branch for different suite. For instance:
- [gce-1.7-master-upgrade-master](https://k8s-testgrid.appspot.com/master-upgrade#gce-1.7-master-upgrade-master&width=20&sort-by-failures=) runs test cases in 1.7  branch after upgrade.
- [gce-1.7-master-upgrade-cluster-new](https://k8s-testgrid.appspot.com/master-upgrade#gce-1.7-master-upgrade-cluster-new&width=20&sort-by-failures=) runs test cases in master  branch after upgrade.

Depends on their configuration in https://github.com/kubernetes/test-infra/blob/master/jobs/config.json.

**Review by MrHohn — COMMENTED** [review]
Technically I think you could send manual cherrypicks (includes only the cluster ip change) directly against 1.7 and 1.6 without changing this value in 1.8 (which seems pointless)?

**Comment by MrHohn** [comment]
cc @bowei

File: test/e2e/network/dns.go [hunk 1]
```diff
@@ -449,7 +449,7 @@ var _ = SIGDescribe("DNS", func() {
 		By("changing the service to type=ClusterIP")
 		_, err = framework.UpdateService(f.ClientSet, f.Namespace.Name, serviceName, func(s *v1.Service) {
 			s.Spec.Type = v1.ServiceTypeClusterIP
-			s.Spec.ClusterIP = "10.0.0.123"
+			s.Spec.ClusterIP = "10.0.0.111"
```
Comment (bowei) [inline test/e2e/network/dns.go]: There needs to at least be a TODO comment to say that this is hardcoded and must lie within the service range.

Why is this changed from x.x.x.123 to .111?

**Comment by MrHohn** [comment]
To speed things up, I sent https://github.com/kubernetes/kubernetes/pull/52804 ahead to fix the invalid cluster IP for 1.7.

File: test/e2e/network/dns.go (hunk 1 above)
Comment (xiangpengzhao) [reply test/e2e/network/dns.go]: I didn't know we can partially cherry-pick a PR when I created this PR. I was just thinking to create a seed PR for chrry-picking, so I change the ClusterIP to be another randomly valid IP :)

**Comment by xiangpengzhao** [comment]
/close
in favor of #52804, thanks @MrHohn!

**Comment by MrHohn** [comment]
@xiangpengzhao Thanks for the ack :)
